modules/mauth-authenticator-akka-http/src/test/scala/com/mdsol/mauth/akka/http/MAuthDirectivesSpec.scala

```diff
@@ -139,7 +178,205 @@ class MAuthDirectivesSpec extends WordSpec with Matchers with ScalatestRouteTest
+
+      Get().withHeaders(
+        RawHeader(MAuthRequest.MCC_TIME_HEADER_NAME, (timeHeader - requestValidationTimeout.toSeconds - 10).toString),
+        RawHeader(MAuthRequest.MCC_AUTHENTICATION_HEADER_NAME, authHeaderV2),
+        RawHeader(MAuthRequest.X_MWS_TIME_HEADER_NAME, (timeHeader - requestValidationTimeout.toSeconds - 10).toString),
+        RawHeader(MAuthRequest.X_MWS_AUTHENTICATION_HEADER_NAME, authHeader)
+      ) ~> route ~> check {
+        inside(rejection) { case MdsolAuthFailedRejection ⇒ }
+      }
+    }
+
+    "reject if public key cannot be found" in {
+      (client.getPublicKey _).expects(appUuid).returns(Future(None))
+      //noinspection ConvertibleToMethodValue
+      (mockEpochTimeProvider.inSeconds _: () => Long).expects().returns(timeHeader)
+
+      Get().withHeaders(
+        RawHeader(MAuthRequest.MCC_TIME_HEADER_NAME, timeHeader.toString),
+        RawHeader(MAuthRequest.MCC_AUTHENTICATION_HEADER_NAME, authHeaderV2),
+        RawHeader(MAuthRequest.X_MWS_TIME_HEADER_NAME, timeHeader.toString),
+        RawHeader(MAuthRequest.X_MWS_AUTHENTICATION_HEADER_NAME, authHeader)
+      ) ~> route ~> check {
+        inside(rejection) { case MdsolAuthFailedRejection ⇒ }
+      }
+    }
+
+    "reject if Authentication header is missing" in {
+      (client.getPublicKey _).expects(appUuid).never
+
+      Get().withHeaders(RawHeader(MAuthRequest.X_MWS_TIME_HEADER_NAME, timeHeader.toString)) ~> route ~> check {
+        inside(rejection) {
+          case MissingHeaderRejection(headerName) ⇒ headerName.replaceAll("_", "-").toLowerCase shouldEqual `MCC-Authentication`.name
+        }
+      }
+    }
+
+    "reject if Time header is missing" in {
+      (client.getPublicKey _).expects(appUuid).never
+
+      Get().withHeaders(RawHeader(MAuthRequest.MCC_AUTHENTICATION_HEADER_NAME, authHeaderV2)) ~> route ~> check {
+        inside(rejection) {
+          case MissingHeaderRejection(headerName) ⇒ headerName.replaceAll("_", "-").toLowerCase shouldEqual `MCC-Time`.name
+        }
+      }
+    }
+  }
+
+  "extractLatestTimeHeader" should {
+    lazy val route =
+      extractLatestTimeHeaderDirective(false) { x ⇒
+        complete(x.toString)
+      }
+
+    "extract time from request" in {
+      Get().withHeaders(RawHeader(MAuthRequest.MCC_TIME_HEADER_NAME, "1234567")) ~> route ~> check {
+        responseAs[String] shouldBe "1234567"
+      }
+    }
+
+    "reject with a MalformedHeaderRejection if supplied with bad format" in {
+      Get().withHeaders(RawHeader(MAuthRequest.X_MWS_TIME_HEADER_NAME, "xyz")) ~> route ~> check {
+        inside(rejection) { case MalformedHeaderRejection("x-mws-time", "x-mws-time header supplied with bad format: [xyz]", None) ⇒ }
+      }
+    }
+
+    "reject with a MalformedHeaderRejection if supplied with bad format for V2" in {
+      Get().withHeaders(
+        RawHeader(MAuthRequest.MCC_TIME_HEADER_NAME, "xyz"),
+        RawHeader(MAuthRequest.X_MWS_TIME_HEADER_NAME, "1234567")
```

Comment:
why is this passed in this test?

Reply:
The client always pick up V2 header MCC_TIME_HEADER_NAME if the V2 header is provided, the time header should be number, otherwise it throws MalformedHeaderRejection

Reply:
I mean, this test will pass if X_MWS_TIME_HEADER_NAME is not passed, so, why passing it?

Reply:
We like to check the client will read and check the V2 header MCC_TIME_HEADER_NAME only, ignore the V1 header passed if MCC_TIME_HEADER_NAME is passed.